### ISSUE TITLE
Document breaking changes in Pelican 4.5

### ISSUE THREAD

**Issue by shniubobo** [getpelican/pelican#2805]
- [x] I have read the [Filing Issues](https://docs.getpelican.com/en/latest/contribute.html#filing-issues) and subsequent “How to Get Help” sections of the documentation.
- [x] I have searched the [issues](https://github.com/getpelican/pelican/issues?q=is%3Aissue) (including closed ones) and believe that this is not a duplicate.

- **OS version and name**: Ubuntu 18.04 (WSL)
- **Python version**: 3.8.6

# Issues

## The first one

**As of https://github.com/getpelican/pelican/commit/58edad6897931f21c7c2f4314cfab33c2b167680, `import pelican.signals` is no longer valid.**

```
$ git reset 58edad^ --hard
$ python -c "import pelican.signals"  # exits 0

$ git reset 58edad --hard
$ python -c "import pelican.signals"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pelican.signals'
```

## The second one

**As of https://github.com/getpelican/pelican/commit/ed1eca160e80bebbfd5870afe1f2a111f388e0ed, legacy plugins with `.` in their names stop working.**

### To reproduce

Create these files:
```
├── content
│   └── 1.rst
├── pelicanconf.py
├── plugins
│   └── m
│       └── test.py
```

```rst
.. content/1.rst
###
123
###
```

```python
# pelicanconf.py
PLUGINS = [
    'm.test',
]
PLUGIN_PATHS = [
    'plugins'
]
```

```python
# plugins/m/test.py
from pelican import signals


def takes_nothing():
    pass


def register():
    signals.initialized.connect(takes_nothing)
```

And then:
```
$ git reset ed1eca^ --hard
$ python -m pelican content
WARNING: Feeds generated without SITEURL set properly may not be valid
WARNING: No timezone information specified in the settings. Assuming your timezone is UTC for feed generation. Check http://docs.getpelican.com/en/latest/settings.html#timezone for more information
CRITICAL: takes_nothing() takes 0 positional arguments but 1 was given
# plugin correctly loaded

$ git reset ed1eca --hard
$ python -m pelican content
WARNING: Feeds generated without SITEURL set properly may not be valid
WARNING: No timezone information specified in the settings. Assuming your timezone is UTC for feed generation. Check http://docs.getpelican.com/en/latest/settings.html#timezone for more information
ERROR: Cannot load plugin `m.test`
  | No module named 'm'
ERROR: Skipping /[path redacted]/content/1.rst: could not find information about 'date'
Done: Processed 0 articles, 0 drafts, 0 pages, 0 hidden pages and 0 draft pages in 0.08 seconds.
# plugin not loaded
```

## Projects affected by these changes

An example: https://github.com/mosra/m.css

And a related issue: https://github.com/mosra/m.css/issues/178

**Comment by avaris** [comment]
First one was [known](https://github.com/getpelican/pelican/pull/2644#issuecomment-548304256), I tried to keep it backwards compatible at first by adding a dummy file but then opted to remove that since no plugin was using that style. `from pelican import signals` will still work.

I'll check second one. Does it still error if you include a `__init__.py` in the `m` folder?

**Comment by shniubobo** [comment]
> First one was [known](https://github.com/getpelican/pelican/pull/2644#issuecomment-548304256), I tried to keep it backwards compatible at first by adding a dummy file but then opted to remove that since no plugin was using that style. `from pelican import signals` will still work.
> 
> I'll check second one. Does it still error if you include an empty `__init__.py` in the `m` folder?

Yes it still errors.

In the affected project mentioned above, there is [an empty `__init__.py`](https://github.com/mosra/m.css/blob/master/plugins/m/__init__.py).

**Comment by avaris** [comment]
This should be enough to fix the second issue, if you want to test it out:

```diff
diff --git a/pelican/plugins/_utils.py b/pelican/plugins/_utils.py
index 78c19e54..a2930551 100644
--- a/pelican/plugins/_utils.py
+++ b/pelican/plugins/_utils.py
@@ -40,6 +40,11 @@ def list_plugins(ns_pkg=None):


 def load_legacy_plugin(plugin, plugin_paths):
+    if '.' in plugin:
+        # it is in a package, so resolve packages first
+        package, _, _ = plugin.rpartition('.')
+        load_legacy_plugin(package, plugin_paths)
+
     # Try to find plugin in PLUGIN_PATHS
     spec = importlib.machinery.PathFinder.find_spec(plugin, plugin_paths)
     if spec is None:
```

I'll make a PR when I have time to add tests for it.

**Comment by shniubobo** [comment]
Thanks for the quick fix! The plugins work fine with this patch.

But what about the first issue? Maybe one extra line in the changelog could save lots of plugin developers' time. You know, a lot of plugins are not included in the pelican-plugins repo (like the affected project I mentioned). These plugins' developers, if they were using `import pelican.signals`, a `BREAKING CHANGES` section in the changelog could help them fix their broken plugins easily. As is in the related issue I mentioned, sometimes this simple import error can be difficult to debug.

**Comment by avaris** [comment]
We can add the changelog entry for the next version about the change.

I could also restore `pelican.signals` but only with an `ImportError` stating that it is deprecated, and tell plugins to switch. That would look like this:

```console
ERROR: Cannot load plugin `m.test`
  | Importing from `pelican.signals` is deprecated. Use `from pelican import signals` or `import pelican.plugins.signals` instead.
```

**Comment by justinmayer** [comment]
I took the liberty of updating the [Pelican 4.5 release post](https://blog.getpelican.com/pelican-4.5-released.html), adding a fourth caveat to the **“Upgrading from previous releases”** section and including the above deprecation details.

**Comment by mosra** [comment]
> This should be enough to fix the second issue

@avaris this is great, thank you :heart: Would significantly simplify everything for all m.css users :)

**Comment by egberts** [comment]
I'm running Pelican 4.5.0 and am still experiencing load_legacy() not found issue.
```console
$ python3.8 -c "import pelican;print(pelican.__version__)"
4.5.0
```

I've done a `git submodule` checkout of m.css into the working area of a website such that this directory layout is:
```
+ egbert.net
|-- content
|-- output
|-- plugins
|-- m.css
   |
   +-- plugins
      +-- ansilexer.py
      +-- dot2svg.py
      +-- latex2svg.py
      +-- latex2svgextra.py
      +-- m
         +-- htmlsanity.py
         +-- metadata.py 
...
```

I used the following plugin settings:
```
PLUGIN_PATHS = [ './plugins', 'm.css/plugins' ]
PLUGINS = [
#          'i18n_subsites',
          'collate_content',
          'sitemap',
          'dateish',
          'tag_cloud',
          'm.alias',
          'm.htmlsanity',
          'm.images',
          'm.math',
          'just_table',
          'code_include',
]
```

And my console output for Pelican was:
```console
ERROR: Cannot load plugin `m.alias`
  | No module named 'm'
ERROR: Cannot load plugin `m.htmlsanity`
  | No module named 'm'
ERROR: Cannot load plugin `m.images`
  | No module named 'm'
ERROR: Cannot load plugin `m.math`
  | No module named 'm'
```
Surely, I must be misreading the documentation or something, am I?

**Comment by avaris** [comment]
> I'm running Pelican 4.5.0 and am still experiencing load_legacy() not found issue.

Still? It's not fixed yet.

**Comment by egberts** [comment]
Yeah, latest master.  Will troubleshoot this.

**Comment by egberts** [comment]
Current starting point for debugging legacy plugin:
```console
(Pdb) b /usr/local/lib/python3.8/dist-packages/pelican-4.5.0-py3.8.egg/pelican/plugins/_utils.py:91
```

**Comment by avaris** [comment]
Not sure what you're trying to debug. It's not fixed. Not even in master, that's why this issue is still open. I didn't have time to make a proper PR yet.

**Comment by egberts** [comment]
Yeah...  turns out there's three ways to do a plugins:

https://packaging.python.org/guides/creating-and-discovering-plugins/

**Comment by justinmayer** [comment]
Thanks to @avaris working on the above-linked PR, this issue should now be addressed. Feel free to test latest master and post a comment here if any follow-up changes are deemed to be warranted.

**Comment by justinmayer** [comment]
Fix for this issue is included in the just-released [Pelican 4.5.1](https://github.com/getpelican/pelican/releases/tag/4.5.1). ✨
